File: python/c21_1_primes_and_factors.py
import unittest
from itertools import groupby
from math import prod,isqrt,log
import sys

sys.setrecursionlimit(max(sys.getrecursionlimit(),1<<20))

def prime(n):
    if n<2:
        return False
    for x in range(2,isqrt(n)+1):
        if n%x==0:
            return False
    return True

def factors(n):
    x=2
    while x*x<=n:
        while n%x==0:
            yield x
            n//=x
        x+=1
    if n>1:
        yield n

def factors_e(n):
    for p,g in groupby(factors(n)):
        yield p,sum(1 for _ in g)

def factor_count(n):
    return prod(a+1 for _,a in factors_e(n))

def factor_sum(n):
    return prod((p**(a+1)-1)//(p-1) for p,a in factors_e(n))

def factor_product(n):
    c=factor_count(n)
    return n**(c//2)*(isqrt(n) if c%2 else 1)

def is_perfect(n):
    return n==factor_sum(n)-n

def density(n):
    return n/log(n)

def eratosthenes(n):
    sieve=[1,1]+[0]*(n-1)
    for x in range(2,n+1):
        if sieve[x]:
            continue
        for u in range(2*x,n+1,x):
            sieve[u]=x
    return sieve

def gcd(a,b):
    if b==0:
        return a
    return gcd(b,a%b)

def totient(n):
    return prod((p-1)*p**(a-1) for p,a in factors_e(n))

class Test(unittest.TestCase):
    def test(self):
        self.assertFalse(prime(1))
        self.assertFalse(prime(35))
        self.assertTrue(prime(7))
        self.assertTrue(prime(19))
        self.assertTrue(prime(41))
        self.assertCountEqual(factors(1),[])
        self.assertCountEqual(factors(84),[2,2,3,7])
        self.assertCountEqual(factors_e(1),[])
        self.assertCountEqual(factors_e(84),[(2,2),(3,1),(7,1)])
        self.assertEqual(factor_count(1),1)
        self.assertEqual(factor_count(84),12)
        self.assertEqual(factor_sum(84),224)
        self.assertEqual(factor_product(84),351298031616)
        self.assertFalse(is_perfect(1))
        self.assertTrue(is_perfect(28))
        self.assertAlmostEqual(density(1000000),72382,0)
        self.assertEqual(eratosthenes(20),[1,1,0,0,2,0,3,0,2,3,5,0,3,0,7,5,2,0,3,0,5])
        self.assertEqual(gcd(24,36),12)
        self.assertEqual(gcd(13,8),1)
        self.assertEqual(totient(12),4)
        self.assertEqual(totient(2),1)

if __name__=='__main__':
    unittest.main()

File: c++/c21_1_primes_and_factors.cpp
#include <bits/stdc++.h>
#include <gtest/gtest.h>
using namespace std;
using ll = long long;
#include "c21_1_primes_and_factors.h"
#include "util.h"

namespace {
	bool prime(ll n) {
		if (n < 2) return false;
		for (ll x = 2; x * x <= n; ++x) {
			if (n % x == 0) return false;
		}
		return true;
	}

	vector<ll> factors(ll n) {
		vector<ll> f;
		for (ll x = 2; x * x <= n; ++x) {
			while (n % x == 0) {
				f.push_back(x);
				n /= x;
			}
		}
		if (n > 1) f.push_back(n);
		return f;
	}

	vector<pair<ll, ll>> factors_e(ll n) {
		vector<pair<ll, ll>> ret;
		for (const auto& v : factors(n) | views::chunk_by(equal_to()))
			ret.emplace_back(v[0], ssize(v));
		return ret;
	}

	ll factor_count(ll n) {
		ll ret = 1;
		for (auto [_, a] : factors_e(n))
			ret *= a + 1;
		return ret;
	}

	ll factor_sum(ll n) {
		ll ret = 1;
		for (auto [p, a] : factors_e(n))
			ret *= (ipow(p, a + 1) - 1) / (p - 1);
		return ret;
	}

	ll factor_product(ll n) {
		ll c = factor_count(n);
		return ipow(n, c / 2) * ((c % 2) ? isqrt(n) : 1);
	}

	bool is_perfect(ll n) {
		return n == factor_sum(n) - n;
	}

	double density(ll n) {
		return n / log(n);
	}

	vector<ll> eratosthenes(ll n) {
		vector<ll> sieve(n + 1);
		sieve[0] = sieve[1] = 1;
		for (int x = 2; x <= n; x++) {
			if (sieve[x]) continue;
			for (int u = 2 * x; u <= n; u += x)
				sieve[u] = x;
		}
		return sieve;
	}

	ll gcd_(ll a, ll b) {
		if (b == 0) return a;
		return gcd_(b, a % b);
	}
}

ll totient(ll n) {
	ll ret = 1;
	for (auto [p, a] : factors_e(n))
		ret *= (p - 1) * ipow(p, a - 1);
	return ret;
}

TEST(C211PrimesAndFactors, prime) {
	EXPECT_FALSE(prime(1));
	EXPECT_FALSE(prime(35));
	EXPECT_TRUE(prime(7));
	EXPECT_TRUE(prime(19));
	EXPECT_TRUE(prime(41));
}

TEST(C211PrimesAndFactors, factors) {
	EXPECT_EQ(factors(1) | ranges::to<multiset>(), multiset<ll>{});
	EXPECT_EQ(factors(84) | ranges::to<multiset>(), (multiset<ll>{2, 2, 3, 7}));
}

TEST(C211PrimesAndFactors, factorsE) {
	EXPECT_EQ(factors_e(1) | ranges::to<multiset>(), (multiset<pair<ll, ll>>{}));
	EXPECT_EQ(factors_e(84) | ranges::to<multiset>(), (multiset<pair<ll, ll>>{{2, 2}, {3,1}, {7,1}}));
}

TEST(C211PrimesAndFactors, factorCount) {
	EXPECT_EQ(factor_count(1), 1);
	EXPECT_EQ(factor_count(84), 12);
}

TEST(C211PrimesAndFactors, factorSum) {
	EXPECT_EQ(factor_sum(84), 224);
}

TEST(C211PrimesAndFactors, factorProduct) {
	EXPECT_EQ(factor_product(84), 351298031616);
}

TEST(C211PrimesAndFactors, isPerfect) {
	EXPECT_FALSE(is_perfect(1));
	EXPECT_TRUE(is_perfect(28));
}

TEST(C211PrimesAndFactors, density) {
	EXPECT_NEAR(density(1'000'000), 72382, 0.5);
}

TEST(C211PrimesAndFactors, eratosthenes) {
	EXPECT_EQ(eratosthenes(20), (vector<ll>{1,1,0,0,2,0,3,0,2,3,5,0,3,0,7,5,2,0,3,0,5}));
}

TEST(C211PrimesAndFactors, gcd) {
	EXPECT_EQ(gcd_(24, 36), 12);
	EXPECT_EQ(gcd_(13, 8), 1);
}

TEST(C211PrimesAndFactors, totient) {
	EXPECT_EQ(totient(12), 4);
	EXPECT_EQ(totient(2), 1);
}
